### ISSUE TITLE
Synology Docker bash terminal fails

### ISSUE THREAD

**Issue by cgradone** [jlesage/docker-crashplan-pro#50]
I noticed this on the previous releases as well but didn't submit. Within the Synology Docker UI, an error is thrown when trying to create a terminal/bash session in the container. The error is: rpc error: code = 2 desc = oci runtime error: exec failed: container_linus.go:247: starting container process caused "exec: \"/bin/bash\": stat /bin/bash: no such file or directory. 

The image of the error message is attached.

![syno-docker err](https://user-images.githubusercontent.com/22548112/35824925-63359290-0a82-11e8-88d7-bb7f6d0a0d81.png)

**Comment by jlesage** [comment]
It’s because bash is not part of the image.
Launching shell with /bin/sh would work.  However, I don’t know if it’s possible to configure that on Synology...

**Comment by cgradone** [comment]
I don't think this can be configured on the Synology. As a reference, this did work with the patters Crashplan distro on Synology.  Would a symbolic link inside the container to /bin/bash work? Not sure where that link would need to be created.

**Comment by PatGitMaster** [comment]
>> It’s because bash is not part of the image.
>> Launching shell with /bin/sh would work. However, I don’t know if it’s possible to configure that on Synology...

Using command sh works just fine on Synology.

**Comment by cgradone** [comment]
The default in Synology Docker UI, when click on Create, is bash. However, running sh as you mentioned works fine. Thanks.

**Comment by bigtfromaz** [comment]
Ok.  This container works just fine on the Synology.  I chose to use the GUI and it’s very simple but there is one constraint and one quirk you need to know.

Here are the steps from start, to running, to upgrading:

Start:


  1.  Open the Docker configuration tool
  2.  Select the Registry tab, search for and download jlesage/crashplan-pro. Choose the “latest” image
  3.  Go to the Image tab, select the crashplan-pro image and select Launch
  4.  I set my CPU priority to Medium and memory limit to about to 1.5 MB
  5.  Go into the Advanced tab, then the Volume tab and then Add folder
     *   Here is where you find the constraint that I mentioned. You will see the first level children of volume1, not volume1 itself.  My personal opinion is that this is a lot better than mounting volume1 directly to the container.  volume1 is fully under the control of the Synology’s configuration system.  It’s owned by root and it’s theirs.  They will change it if they see fit.
     *   Just add each folder that you want Crashplan to see.  Mount them all under /storage; e.g. My “Idrive” folder is mounted to /storage/I.  Note that I chose a different mount point name than the original name of the folder.  This allows me to relocate Idrive to another volume on the Synology, and remount it to the same mount point.  CrashPlan will see no change so resync will not occur.
     *   Don’t forget to define the “config” mount point and set it to read/write.
  6.  You can ignore the Port Settings and Links tabs.
  7.  On the Environment tab, follow jlesage’s notes.
     *   However I decided to set the USER_ID to my ID number (use the id command from a console to get your ID number.  I set the GROUP_ID to 101, the default ID for the Synology ADMIN group.  Running as root is always a bad idea.  This avoids running as root while providing CrashPlan all the permissions it needs.
     *   If you worry about the CrashPlan app’s behavior, you can set your mount points to read only with the exception of the config mount point.
     *   I use high-res monitors so I set the DISPLAY WIDTH and HEIGHT to something more appropriate for a 4k monitor.
  8.  Finally, make sure the Network tab is set appropriately for your network. Most networks should be fine using a “host” network.

Start it up and check it out.  Logs are in the config folder.  The first run may take a while to resync your new mappings but it goes fast and if the files are already on Crashplan, they won’t be resent.  Note the terminal window into the container does not work from the GUI.  I haven’t tried it from a command line.  I actually have never felt the need to use it.

Upgrades:

Upgrades are a really easy.

  1.  Select the container and stop it
  2.  Go to the Settings pulldown and export the container’s settings.  Put them in a folder on your Synology, or you can download them to your local machine.  The settings file for the container is a JSON file.  The GUI can use it to recreate your container.
  3.  Delete the container.
  4.  Select the Registry tab, search for and download jlesage/crashplan-pro, choose the “latest image”.
  5.  You will see a little marker on the existing image that tells you it is downloading.
  6.  When the download is finished, Go to the Container tab and import the container using the settings file you exported.
  7.  Start the container and it’ll start right back up.

Finally, there is a quirk with Synology regarding file watching. The article jlesage pints us to at CODE42 is accurate however the Synology keeps its configuration in /etc.defaults/sysctl.conf .  This is not the usual place to find sysctl.conf.
So, add this line:
fs.inotify.max_user_watches=1048576
to /etc.defaults/sysctl.conf

IMPORTANT: Make a note that you need to check /etc.defaults/sysctl.conf after installing DSM upgrades.  Sometimes the rewrite the file wiping out your change.  Just read the line and restart the Synology.

Follow this plan and it should work.

Cheers

**Comment by jlesage** [comment]
@bigtfromaz, the only comment I have with what you said is that I would recommend using a "bridge" network type.  It's the default Docker network and it offers better security by isolating the container from the host.  Thus, internal processes that are listening on 127.0.0.1 are not visible by the host.
Also, a "bridge" network makes possible to run multiple containers at the same time even if they use the same ports internally.

**Comment by bigtfromaz** [comment]
Point taken and it’s a good one.

I haven’t researched the Synology setup’s effect on routing the container’s outbound requests for DNS and gateway services.  I started there using some defaults, but the way  Synology set it up, the container couldn’t reach the internet.  I saw enough to know that it put the container in a different subnet.   In the interest of time I switched to Host mode and figured I would sort it out later.  But…I haven’t gotten around to it.  Maybe I’ll play around with it tonight.
